aiida/orm/nodes/data/singlefile.py

```diff
@@ -26,15 +26,21 @@ class SinglefileData(Data):
 
     DEFAULT_FILENAME = 'file.txt'
 
-    def __init__(self, file, **kwargs):
+    def __init__(self, file, filename=None, **kwargs):
         """Construct a new instance and set the contents to that of the file.
 
-        :param file: an absolute filepath or filelike object whose contents to copy
+        :param file: an absolute filepath or filelike object whose contents to copy.
+            Hint: Pass io.BytesIO(b"my string") to construct the SinglefileData directly from a string.
+        :param filename: specify filename to use (defaults to name of provided file).
         """
         # pylint: disable=redefined-builtin
         super(SinglefileData, self).__init__(**kwargs)
         if file is not None:
-            self.set_file(file)
+            if filename is None:
+                # for backwards compatibility don't assume that set_file has a 'filename' argument
```

Comment:
I don't think this accomplishes what you intend it to do. What if the sub class does _not_ override the constructor and a user passes `filename`, it will then go to line 43 and except on `set_file` if the overridden version does not add `filename`. Doing the switch on the value to guard against the signature seems fragile. Shouldn't we just call with the filename argument and catch `TypeError`. Even that is non-ideal, but the only remaining option would be to inspect the signature of the method which is also fragile I think

Reply:
>  What if the sub class does not override the constructor and a user passes filename

I would consider it ok if an exception is raised in this case - since the user is using the class in a way it is not designed by the author.
What is not ok is to assume that `set_file` supports the `filename` argument as there may be many subclasses of `SinglefileData` out there that don't.

> Shouldn't we just call with the filename argument and catch TypeError. Even that is non-ideal, but the only remaining option would be to inspect the signature of the method which is also fragile I think

Hm... I would prefer the guard based on the argument over catching any `TypeError` thrown by the function call.

We could do inspection of the function arguments (`if 'filename' not in inspect.getargspec(self.f)[0]`) but what do you if we see that `filename` is missing?
If it is missing, then perhaps the plugin author should update it - we could raise a warning.
But we would still need to check the user-provided value - if it is not `None`, we cannot just silently ignore it.

Reply:
Ok fine to keep it. Is there a path that we can remove this clause in a future version though for simplification? I realize a "deprecation" warning does not really work here. So how would we go about warning people that the new `set_file` needs a new argument?

Reply:
hm, I think we can add the deprecation warning (we just need to keep the logic based on the value of the filename argument as well).
done in the last commit.